### PR TITLE
Implement `SubscribeCommitteeSubnet` method

### PR DESCRIPTION
**Pull request by terencechain**
This PR implements `SubscribeCommitteeSubnet` for BN RPC server and uses it up the VC client. It opens up a few open ended question that should be explore in the following PR. How does a node subscribe differently for aggregaor and attester? 

Change list:
* Change API of committee ID cache to be `slot` based instead of `epoch` based. Using epoch as key doesn't make sense to aggregate
* Change the updating of committeeID cache on per `SubscribeCommitteeSubnet` request instead of `GetAttestationData` and `ProposeAttestation` requests
* Added round trip tests for committee ID cache, it has 0 test

cc @rauljordan @nisdas

### CODE REVIEW

File: beacon-chain/rpc/validator/attester.go [hunk 1]
```diff
@@ -235,3 +224,13 @@ func (vs *Server) waitToOneThird(ctx context.Context, slot uint64) {
 		}
 	}
 }
+
+// SubscribeCommitteeSubnet subscribes to the committee ID subnet given subscribe request.
+func (vs *Server) SubscribeCommitteeSubnet(ctx context.Context, req *ethpb.CommitteeSubnetSubscribeRequest) (*ptypes.Empty, error) {
+	if req.IsAggregator {
+		cache.CommitteeIDs.AddID(req.Slot, req.CommitteeId)
+	} else {
+		// What should be done for an attester?
```
Comment (prestonvanloon) [inline beacon-chain/rpc/validator/attester.go:233]: What is this?

**Comment by terencechain** [comment]
Marking this a blocked until I get feedback on caching committee ID values for `attester`

File: beacon-chain/cache/committee_ids.go [hunk 2]
```diff
@@ -16,27 +17,30 @@ type committeeIDs struct {
...
 	if err != nil {
 		panic(err)
 	}
 	return &committeeIDs{cache: cache}
 }
 
-// AddIDs to the cache for attestation committees by epoch.
-func (t *committeeIDs) AddIDs(indices []uint64, epoch uint64) {
+// AddID adds committee ID for subscribing subnet for the attester and/or aggregator of a given slot.
+func (t *committeeIDs) AddID(slot uint64, committeeID uint64) {
```
Comment (nisdas) [inline beacon-chain/cache/committee_ids.go:29]: Not sure this is the correct way for this , we restrict this to only 1 epoch. We need an epoch lookahead for our validators.

File: beacon-chain/p2p/service.go [hunk 3]
```diff
@@ -348,13 +348,13 @@ func (s *Service) MetadataSeq() uint64 {
...
 // to be dynamically discoverable by others given our tracked committee id's.
-func (s *Service) RefreshENR(epoch uint64) {
+func (s *Service) RefreshENR(slot uint64) {
 	// return early if discv5 isnt running
 	if s.dv5Listener == nil {
 		return
 	}
 	bitV := bitfield.NewBitvector64()
-	committees := cache.CommitteeIDs.GetIDs(epoch)
+	committees := cache.CommitteeIDs.GetAttesterCommitteeIDs(slot)
```
Comment (nisdas) [inline beacon-chain/p2p/service.go:357]: we will need to add both attester + aggregator indexes

File: beacon-chain/sync/subscriber_committee_index_beacon_attestation.go [hunk 4]
```diff
@@ -51,6 +51,6 @@ func (r *Service) committeesCount() int {
 
 func (r *Service) committeeIndices() []uint64 {
 	currentEpoch := helpers.SlotToEpoch(r.chain.HeadSlot())
-	return sliceutil.UnionUint64(cache.CommitteeIDs.GetIDs(currentEpoch),
-		cache.CommitteeIDs.GetIDs(currentEpoch+1))
+	return sliceutil.UnionUint64(cache.CommitteeIDs.GetAttesterCommitteeIDs(currentEpoch),
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/sync/subscriber_committee_index_beacon_attestation.go]: ah i see, not sure if using head slot there was correct to begin with, i changed it to use `chain.CurrentSlot()`

File: beacon-chain/p2p/service.go (hunk 3 above)
Comment (terencechain) [reply beacon-chain/p2p/service.go]: thanks!

File: beacon-chain/p2p/service.go [hunk 5]
```diff
@@ -348,13 +349,14 @@ func (s *Service) MetadataSeq() uint64 {
...
 // to be dynamically discoverable by others given our tracked committee id's.
-func (s *Service) RefreshENR(epoch uint64) {
+func (s *Service) RefreshENR(slot uint64) {
 	// return early if discv5 isnt running
 	if s.dv5Listener == nil {
 		return
 	}
 	bitV := bitfield.NewBitvector64()
-	committees := cache.CommitteeIDs.GetIDs(epoch)
+	committees := sliceutil.UnionUint64(cache.CommitteeIDs.GetAttesterCommitteeIDs(slot),
```
Comment (nisdas) [inline beacon-chain/p2p/service.go:358]: this isnt accurate, we still refresh by epoch. What we need to do here is loop through all the slots in the epoch and aggregate the relevant committe indexes

**Pull request merged**